corporate/views/remote_billing_page.py

```diff
@@ -99,7 +110,72 @@ def remote_realm_billing_finalize_login(
     except signing.BadSignature:
         raise JsonableError(_("Invalid billing access token."))
 
+    # Now we want to fetch (or create) the RemoteRealmBillingUser object implied
+    # by the IdentityDict. We'll use this:
+    # (1) If the user came here via just GET, we want to show them a confirmation
+    #     page with the relevant info details before finalizing login. If they wish
+    #     to proceed, they'll approve the form, causing a POST, bring us to case (2).
+    # (2) If the user came here via POST, we finalize login, using the info from the
+    #     IdentityDict to update the RemoteRealmBillingUser object if needed.
     remote_realm_uuid = identity_dict["remote_realm_uuid"]
+    remote_server_uuid = identity_dict["remote_server_uuid"]
+    try:
+        remote_server = get_remote_server_by_uuid(remote_server_uuid)
+        remote_realm = RemoteRealm.objects.get(uuid=remote_realm_uuid, server=remote_server)
+    except ObjectDoesNotExist:
+        # These should definitely still exist, since the access token was signed
+        # pretty recently. (And we generally don't delete these at all.)
+        raise AssertionError
+
+    user_dict = identity_dict["user"]
+
+    user_email = user_dict["user_email"]
+    user_full_name = user_dict["user_full_name"]
+    user_uuid = user_dict["user_uuid"]
+    remote_user, created = RemoteRealmBillingUser.objects.get_or_create(
+        defaults={"full_name": user_full_name, "email": user_email},
+        remote_realm=remote_realm,
+        user_uuid=user_uuid,
+    )
+    assert (
+        settings.TERMS_OF_SERVICE_VERSION is not None
+    ), "This is only run on the bouncer, which has ToS"
+    tos_consent_needed = int(settings.TERMS_OF_SERVICE_VERSION.split(".")[0]) > int(
+        remote_user.tos_version.split(".")[0]
+    )
+
+    if request.method == "GET":
+        context = {
+            "remote_server_uuid": remote_server_uuid,
+            "remote_realm_uuid": remote_realm_uuid,
+            "remote_realm_host": remote_realm.host,
+            "user_email": user_email,
+            "user_full_name": user_full_name,
+            "tos_consent_needed": tos_consent_needed,
+            "action_url": reverse(
+                remote_realm_billing_finalize_login, args=(signed_billing_access_token,)
+            ),
+        }
+        return render(
+            request,
+            "corporate/remote_realm_billing_finalize_login_confirmation.html",
+            context=context,
+        )
+
+    if request.method != "POST":
+        return HttpResponseNotAllowed(["GET", "POST"])
```

Comment:
Let's put this sort of basic validation at the top, before doing real work.

Reply:
Move to the top